### PR TITLE
Issue 1538: Implement support for generating test artifacts with a specified size for Pypi

### LINKED ISSUE #1538: Implement support for generating test artifacts with a specified size for Pypi
# Task Description

We need to be able to create test artifacts with a specified size.

# Classes of Interest

* `org.carlspring.strongbox.artifact.generator.PypiArtifactGenerator`
* `org.carlspring.strongbox.testing.artifact.PypiTestArtifact`

# Tasks

The following tasks will need to be carried out:
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.artifact.generator.PypiArtifactGenerator`.
* [ ] Apply the necessary changes to the `org.carlspring.strongbox.testing.artifact.PypiTestArtifact`.
* [ ] Add tests for this.

# Task Relationships

This task:
* Is a sub-task of: #1534 
* Relates to: #1535 , #1536 , #1537, #1540

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @fuss86
  * @sbespalov
  * @steve-todorov

### PR DESCRIPTION
**Pull request by m-podhajski**
# Pull Request Description

This pull request closes #1538 .

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@Execution(CONCURRENT)
+@ContextConfiguration(classes = PyPiArtifactGeneratorTestConfig.class)
+public class PyPiArtifactGeneratorTest {
+
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class,
+            ArtifactManagementTestExecutionListener.class })
+    @Test
+    public void testArtifactGeneration(@PypiTestRepository(repositoryId = "repositoryId") Repository repository,
+                                       @PypiTestArtifact(repositoryId = "repositoryId", resource="org-carlspring-123-strongbox-testing-pypi.whl", size = 2048)
+                                               Path artifactPath)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java]: Cosmetic: Indentation: Could you please change it like this:
```
    public void testArtifactGeneration(@PypiTestRepository(repositoryId = "repositoryId") Repository repository,
                                       @PypiTestArtifact(repositoryId = "repositoryId",
                                                         resource="org-carlspring-123-strongbox-testing-pypi.whl",
                                                         size = 2048)
                                       Path artifactPath)
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.nio.file.Path;
+
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+@SpringBootTest
+@ActiveProfiles(profiles = "test")
+@Execution(CONCURRENT)
+@ContextConfiguration(classes = PyPiArtifactGeneratorTestConfig.class)
+public class PyPiArtifactGeneratorTest {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java]: Cosmetic: Braces on new lines, please. :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class,
+            ArtifactManagementTestExecutionListener.class })
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generation/PyPiArtifactGeneratorTest.java]: Cosmetic: Please, change it to be aligned like this:
```
    @ExtendWith({ RepositoryManagementTestExecutionListener.class,
                  ArtifactManagementTestExecutionListener.class })
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 4]
```diff
@@ -100,34 +105,39 @@ public Path generateArtifact(URI uri,
             throws IOException
     {
         PypiArtifactCoordinates coordinates = PypiArtifactCoordinates.parse(FilenameUtils.getName(uri.toString()));
-        return generateArtifact(coordinates);
+        return generateArtifact(coordinates, size);
     }
 
-    public Path generateArtifact(PypiArtifactCoordinates coordinates)
+    public Path generateArtifact(PypiArtifactCoordinates coordinates, long size)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: Could you please rename `size` to `byteSize` here and across the pull request? Thanks! :)

@wpater has already been following this naming convention.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 5]
```diff
@@ -100,34 +105,39 @@ public Path generateArtifact(URI uri,
...
         String packagePath = coordinates.toPath();
 
         Path fullPath = basedir.resolve(packagePath);
 
-        try(ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(fullPath)))
+        new File(fullPath.getParent().toString()).mkdirs();
+        Path path = Paths.get(fullPath.toString());
+        OutputStream outputStream = Files.newOutputStream(path, CREATE);
+
+
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: Cosmetic: Could you please remove this empty line? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 6]
```diff
@@ -179,11 +189,14 @@ private void createSourcePackageFiles(ZipOutputStream zos,
         byte[] sourcesContent = String.format(SOURCES_CONTENT, scriptPath, eggPkgInfoPath, sourcesPath,
                                               dependencyLinksPath, topLevelPath).getBytes();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: Cosmetic: (I know this is not part of your changes), but could you please chop these arguments as one per line like this:
```
        byte[] sourcesContent = String.format(SOURCES_CONTENT,
                                              scriptPath,
                                              eggPkgInfoPath,
                                              sourcesPath,
                                              dependencyLinksPath,
                                              topLevelPath).getBytes();
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        CommonConfig.class,
+        TestingCoreConfig.class,
+        EventsConfig.class,
+        CronTasksConfig.class,
+        StorageCoreConfig.class,
+        StorageApiConfig.class,
+        ClientConfig.class,
+        PypiLayoutProviderConfig.class,
+})
+public class PyPiArtifactGeneratorTestConfig {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java]: Cosmetic: Braces on new lines, please.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+@Import({
+        CommonConfig.class,
+        TestingCoreConfig.class,
+        EventsConfig.class,
+        CronTasksConfig.class,
+        StorageCoreConfig.class,
+        StorageApiConfig.class,
+        ClientConfig.class,
+        PypiLayoutProviderConfig.class,
+})
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java]: Cosmetic: Could you please change it like this:
```
@Import({ CommonConfig.class,
           TestingCoreConfig.class,
           EventsConfig.class,
           CronTasksConfig.class,
           StorageCoreConfig.class,
           StorageApiConfig.class,
           ClientConfig.class,
           PypiLayoutProviderConfig.class })
```

Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java (hunk 7 above)
Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PyPiArtifactGeneratorTestConfig.java]: Also, why do you actually need this class? I think you should drop it. You shouldn't need tese contexts for the artifact generator.

**Comment by carlspring** [comment]
@m-podhajski ,

Thank you for your contribution! Could you please have a look at the 
[coding convention](https://strongbox.github.io/developer-guide/coding-convention.html)?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 9]
```diff
@@ -91,43 +96,48 @@ public Path generateArtifact(String id,
...
+    public Path generateArtifact(PypiArtifactCoordinates coordinates,
+                                 long byteSize)
             throws IOException
     {
         String packagePath = coordinates.toPath();
 
         Path fullPath = basedir.resolve(packagePath);
 
-        try(ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(fullPath)))
+        new File(fullPath.getParent().toString()).mkdirs();
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: Please, use Java NIO.2 when possible:
```suggestion
        Path fullPathParent = fullPath.getParent();
        Files.createDirectories(fullPathParent);
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 10]
```diff
@@ -91,43 +96,48 @@ public Path generateArtifact(String id,
...
+        Path path = Paths.get(fullPath.toString());
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: This variable `path` is the same as `fullPath`. Please remove it.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 11]
```diff
@@ -243,7 +261,31 @@ private void createWheelPackageFiles(ZipOutputStream zos,
...
+
+    private void createZipEntryWithRandom(ZipOutputStream zos,
+                                          String path,
+                                          long byteSize)
+            throws IOException
+    {
+        ZipEntry zipEntry = new ZipEntry(path);
+
+        zos.putNextEntry(zipEntry);
+        RandomInputStream ris = new RandomInputStream(byteSize);
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: Please, use try-with-resources here.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 12]
```diff
@@ -243,7 +261,31 @@ private void createWheelPackageFiles(ZipOutputStream zos,
...
                                               .append(recordPath)
                                               .append(",,");
+
         createZipEntry(zos, recordPath, recordContent.toString().getBytes());
+
+        String randomPath = dirPath + "/" + "RANDOM";
+        createZipEntryWithRandom(zos, randomPath, byteSize);
+    }
+
+    private void createZipEntryWithRandom(ZipOutputStream zos,
```
Comment (wpater) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]: This method is already implemented in `TestFileUtils`

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java [hunk 13]
```diff
@@ -115,6 +115,17 @@ public PypiArtifactCoordinates(String distribution,
...
 
+    public PypiArtifactCoordinates()
+    {
+        this(DISTRIBUTION,
+                VERSION,
+                "",
+                LANGUAGE_IMPLEMENTATION_VERSION,
+                ABI,
+                PLATFORM,
+                WHEEL_EXTENSION);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: Cosmetic: Could you please align chunked paramters to be one per line, aligned by the first parameter like this:
```
doFoo(param1,
      param2,
      param3);
```
For more details, please check our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/config/PypiLayoutProviderConfig.java [hunk 14]
```diff
@@ -60,12 +57,14 @@ public PypiFileSystemProvider pypiFileSystemProvider(FileSystemProvider provider
...
+        LayoutFileSystemProviderFactory providerFactory = pypiRepositoryFileSystemProviderFactory();
 
+        return (repository) -> {
             StorageProvider storageProvider = storageProviderRegistry.getProvider(repository.getImplementation());
 
-            return pypiRepositoryFileSystem(propertiesBooter, repository, storageProvider.getFileSystem(),
+            return pypiRepositoryFileSystem(propertiesBooter,
+                                            repository,
+                                            storageProvider.getFileSystem(),
                                              providerFactory.create(repository));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/config/PypiLayoutProviderConfig.java]: Cosmetic: Could you please re-align the last parameter, as well? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java [hunk 15]
```diff
@@ -115,6 +115,17 @@ public PypiArtifactCoordinates(String distribution,
         setPackaging(packaging);
     }
 
+    public PypiArtifactCoordinates()
+    {
+        this(DISTRIBUTION,
+             VERSION,
+             "",
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: Also, I wanted to ask, why is the third parameter a blank string? Shouldn't it be like this:
```
    public PypiArtifactCoordinates()
    {
        this(DISTRIBUTION,
             VERSION,
             BUILD,
             LANGUAGE_IMPLEMENTATION_VERSION,
             ABI,
             PLATFORM,
             PACKAGING);
    }
```

And another cosmetic change, if you don't mind, could you please move the empty constructor above the other one? It's sort of a convention we have. Thanks! :)

Comment (m-podhajski) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: It should be empty or be a number but BUILD=="build" (not a number :))

Comment (m-podhajski) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: maybe the BUILD variable should be changed?

Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: These constructors look like crap. This is actually not how they should work. I believe the initial implementation, (now that I look at it more closely), is mistaken.

If you have a look at the `MavenArtifactCoordinates`, `NpmArtifactCoordinates`, or `NugetArtifactCoordinates`, you'll see what I mean. Each of the setter methods should actually be invoking `setCoordinate`.

Would you mind fixing this as well? I know it isn't part of your changes at all, but this seems to be quite mistaken and misleading.

Comment (m-podhajski) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates/PypiArtifactCoordinates.java]: ok, no problem :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/PypiArtifactGeneratorStrategy.java [hunk 16]
```diff
@@ -29,6 +29,6 @@ public Path generateArtifact(PypiArtifactGenerator artifactGenerator,
                                                                           "none",
                                                                           "any",
                                                                           (String) attributesMap.get("packaging"));
-        return artifactGenerator.generateArtifact(coordinates);
+        return artifactGenerator.generateArtifact(coordinates, size);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/testing/artifact/PypiArtifactGeneratorStrategy.java]: Cosmetic: Could you please change this to `byteSize`?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 17]
```diff
@@ -176,14 +183,22 @@ private void createSourcePackageFiles(ZipOutputStream zos,
...
-                                              dependencyLinksPath, topLevelPath).getBytes();
+        byte[] sourcesContent = String.format(SOURCES_CONTENT,
+                                              scriptPath,
+                                              eggPkgInfoPath,
+                                              sourcesPath,
+                                              dependencyLinksPath,
+                                              topLevelPath).getBytes();
         createZipEntry(zos, sourcesPath, sourcesContent);
+
+        String randomPath = eggDirectory + "/" + "RANDOM.txt";
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]:
```suggestion
        String randomPath = eggDirectory + "/RANDOM.txt";
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java [hunk 18]
```diff
@@ -243,7 +258,11 @@ private void createWheelPackageFiles(ZipOutputStream zos,
                                                                     topLevelContent.length))
                                               .append(recordPath)
                                               .append(",,");
+
         createZipEntry(zos, recordPath, recordContent.toString().getBytes());
+
+        String randomPath = dirPath + "/" + "RANDOM";
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/artifact/generator/PypiArtifactGenerator.java]:
```suggestion
        String randomPath = dirPath + "/RANDOM";
```

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PypiLayoutProviderTestConfig.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            return !context.getRegistry().containsBeanDefinition("cronTaskConfigurationServiceImpl") &&
+                   !context.getRegistry().containsBeanDefinition("cronJobSchedulerServiceImpl");
+        }
+    }
+
+    @Bean
+    public ArtifactResolutionServiceHelper artifactResolutionServiceHelper()
+    {
+        return new ArtifactResolutionServiceHelper();
+    }
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PypiLayoutProviderTestConfig.java]: This bean method should be placed before the static class, according to [our coding convention](https://strongbox.github.io/developer-guide/coding-convention.html).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PypiLayoutProviderTestConfig.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+          ClientConfig.class })
+public class PypiLayoutProviderTestConfig
+{
+
+    @Bean(name = "mockedConfigurationFileManager")
+    @Primary
+    ConfigurationFileManager configurationFileManager(YAMLMapperFactory yamlMapperFactory)
+            throws IOException
+    {
+        final ConfigurationFileManager configurationFileManager = Mockito.spy(new ConfigurationFileManager(yamlMapperFactory));
```
Comment (ptirador) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/config/PypiLayoutProviderTestConfig.java]: Please, indent this according to [our coding convention](https://strongbox.github.io/developer-guide/coding-convention.html).

**Comment by carlspring** [comment]
Thank you @m-podhajski ! Congratulations on your first pull request! :)

**Pull request merged**